gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -252,4 +263,40 @@ MacroBuiltin::compile_error (Location invoc_locus, AST::MacroInvocData &invoc)
   return AST::ASTFragment::create_error ();
 }
 
+/* Expand builtin macro concat!(), which joins all the literal parameters
+   into a string with no delimiter. */
+
+AST::ASTFragment
+MacroBuiltin::concat (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  auto invoc_token_tree = invoc.get_delim_tok_tree ();
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (std::move (lex));
+  auto str = std::string ();
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+
+  /* NOTE: concat! could accept no argument, so we don't have any checks here */
+  while (parser.peek_current_token ()->get_id () != last_token_id)
+    {
+      auto lit_expr = parser.parse_literal_expr ();
+      if (lit_expr)
+	{
+	  str += lit_expr->as_string ();
+	}
+      else
+	{
+	  rust_error_at (parser.peek_current_token ()->get_locus (),
+			 "argument must be a constant literal");
+	  return AST::ASTFragment::create_error ();
```

Comment:
This is definitely a nitpick, but: Instead of returning early here, we could emit errors as we go, going through the whole loop. This would help in reporting to the user that there might be multiple places in which they're using non-literals in their invocation to `concat!()`. This could definitely be done in a separate PR so no worries about that, but if you want we can open up an issue for it. Then, after the loop, if an error has been seen, return an error `ASTFragment`

Reply:
> This is definitely a nitpick, but: Instead of returning early here, we could emit errors as we go, going through the whole loop. This would help in reporting to the user that there might be multiple places in which they're using non-literals in their invocation to `concat!()`. This could definitely be done in a separate PR so no worries about that, but if you want we can open up an issue for it. Then, after the loop, if an error has been seen, return an error `ASTFragment`

Acknowledged. I thought we need to exit early to avoid reporting false-positives.

Reply:
Yeah this is always a balance i think for now lets try and produce as many errors as possible then we can examine if that becomes a problem down the line.